scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -160,41 +188,112 @@
+    ;       normal exit states (including exit-with-failure). I.e., these MUST only be invoked if expected state is not
+    ;       in a terminal state. These functions return new expected-state-dict's so should be invoked last in almost
+    ;       all cases. (However, in a few cases, (delete-task ? ?) can be invoked afterwards. It deletes the key from
+    ;       the expected state completely. Should only be invoked if we're nil.)
+    ;
+    ;   weird can be called at any time to indicate we're in a wierd state and extra logging is warranted.
+    ;
+    ;   We always end with one of  pod-has-just-completed/pod-was-killed/pod-has-started or delete-task.
+    ;
+    ;   We always update datomic first (with pod-has-* pod-was-*), etc, then we update kubernetes so we can handle restarts.
+    ;
+    ;   (delete-task ? ?) is only valid when kubernetes is in a terminal-state.
+    ;
+    ;   Note that we have release semantics. We could fail to process any operation. This means that, where relevant, we
+    ;     first write to datomic, then change into the next step. We thus show some intermediate states in the machine
+    ;     that seem like they could be skipped.
+    ;
+    ;   In the (:completed,*) states, we delete from datomic, transitioning into (:completed,:missing) and thence to (:missing, :missing)
+    ;     When we're in a terminal pod state.
+    ;
+    ;   We only delete a pod if and only if the pod is in a terminal (or unknown) state.
+    ; In some cases, we may mark mea culpa retries (if e.g., the node got reclaimed, so the pod got killed). In weird cases, we always need to mark mea culpa retries.
     (let
       [new-expected-state-dict (case (vector (or expected-state :missing) (or (:state synthesized-state) :missing))
                                  [:expected/starting :missing] (launch-task api-client expected-state-dict)
                                  [:expected/starting :pod/running] (pod-has-started kcc existing-state-dict)
-                                 ; TODO We need to add cases for the other [:expected/starting *] states in.
-                                 [:expected/starting :pod/waiting] expected-state-dict
-                                 [:expected/starting :pod/succeeded] (pod-has-just-completed kcc existing-state-dict)
-                                 [:expected/starting :pod/failed] (pod-has-just-completed kcc existing-state-dict)
+                                 [:expected/starting :pod/waiting] expected-state-dict ; Its starting. Can be stuck here.
+                                 [:expected/starting :pod/succeeded] (pod-has-just-completed kcc existing-state-dict) ; Finished fast.
+                                 [:expected/starting :pod/failed] (pod-has-just-completed kcc existing-state-dict) ; TODO: May need to mark mea culpa retry
                                  [:expected/running :pod/running] expected-state-dict
```

Comment:
If this expected state is well-known and deterministic, why not hard-code the value?

Reply:
Part of my confusion is that the new state needs to only contain the `:expected-state` key (based on at least `pod-was-killed`). It is not clear to me how the additional metadata you want in the dictionary is being used or needed in the current code.

Reply:
The extra metadata exists in :expected/starting. I expect it to be needed for mea culpa retries to thread the reason until pod-has-completed can write it back.